infrastructure/ssz/src/main/java/tech/pegasys/teku/infrastructure/ssz/tree/SimpleLeafNode.java

```diff
@@ -22,25 +22,39 @@
 
 class SimpleLeafNode implements LeafNode, TreeNode {
 
-  private final byte[] data;
+  private final Bytes data;
+  private Bytes32 hashTreeRoot;
 
   public SimpleLeafNode(Bytes data) {
     checkArgument(data.size() <= MAX_BYTE_SIZE);
-    this.data = data.toArrayUnsafe();
+    if (data.size() == MAX_BYTE_SIZE) {
+      // if data is Bytes32, it will pass throw with no object creation
+      // otherwise translate types to Bytes32
+      this.hashTreeRoot = Bytes32.wrap(data.copy());
+      this.data = hashTreeRoot;
+      return;
+    }
+    // If we store data as is, some Bytes instances (ie ConcatenatedBytes) will
+    // perform worse during serialization. So we want to translate them to a
+    // single array-backed Bytes
+    //
+    // in this case, the following seems to perform better than Bytes.wrap(data.copy())
+    this.data = Bytes.wrap(data.toArrayUnsafe());
   }
 
   @Override
   public Bytes getData() {
-    return Bytes.wrap(data);
+    return data;
   }
 
   @Override
   public Bytes32 hashTreeRoot() {
-    if (data.length == MAX_BYTE_SIZE) {
-      return Bytes32.wrap(data);
-    } else {
-      return Bytes32.wrap(Arrays.copyOf(data, MAX_BYTE_SIZE));
+    if (hashTreeRoot != null) {
+      return hashTreeRoot;
     }
+
+    hashTreeRoot = Bytes32.wrap(Arrays.copyOf(data.toArrayUnsafe(), MAX_BYTE_SIZE));
```

Comment:
i wonder if we should just calculate this on construction anyway... then this is just a getter

Reply:
Yes but I'm a bit concerned by the memory overhead but I just had another idea. Tomorrow I'll explore it

Reply:
Yes but I'm a bit concerned by the memory overhead but I just had another idea. Tomorrow I'll explore it

Reply:
> i wonder if we should just calculate this on construction anyway... then this is just a getter
I'd prefer to have a lazy cache, Anton seems to agree here :)